### PR TITLE
Dropout cleanup

### PR DESCRIPTION
**Pull request by tomkocmi**
Please check it thoroughly, even that test are running I don't want to break NM.

### CODE REVIEW

File: neuralmonkey/decoders/output_projection.py [hunk 1]
```diff
@@ -49,9 +49,8 @@ def _projection(prev_state, prev_output, ctx_tensors):
     return _projection
 
 
-# TODO fix this so it does not need dropout_plc. But it needs to be
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/output_projection.py]: Tohle TODO tu musí zůstat, ten problém se neopravil. Jenom jsi dropout placeholder nahradil train_mode placeholderem. Musí se to vymyslet tak, aby tahle funkce nemusela dostávat žádný tenzory, protože se volá z konfiguráku a tam žádný nejsou.

Comment (jindrahelcl) [inline neuralmonkey/decoders/output_projection.py]: Ideálně by tu nemělo být ani ``dropout_keep_prob``, tj. to TODOčko sem vrať.

File: neuralmonkey/encoders/factored_encoder.py [hunk 2]
```diff
@@ -82,8 +83,7 @@ def _get_birnn_cells(self):
 
     # pylint: disable=too-many-locals
     def _create_encoder_graph(self):
-        self.dropout_placeholder = tf.placeholder(tf.float32, name="dropout")
-        self.is_training = tf.placeholder(tf.bool, name="is_training")
+        self.train_mode = tf.placeholder(tf.bool, name="is_training")
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/factored_encoder.py]: když už to jednotíš, tak by tohle mělo mít `name="train_mode"`.. Sice je to i tak pokaždý jinak, ale `train_mode` je ideální jméno

File: neuralmonkey/nn/mlp.py [hunk 3]
```diff
@@ -6,26 +6,25 @@ class MultilayerPerceptron(object):
     """ General implementation of the multilayer perceptron. """
 
     # pylint: disable=too-many-arguments
-    def __init__(self, mlp_input, layer_configuration, dropout_plc,
+    def __init__(self, mlp_input, layer_configuration, dropout_keep_prob,
```
Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: když už přidáváš anotace, udělej celou funkci a ne jen jeden parametr

File: neuralmonkey/nn/mlp.py [hunk 4]
```diff
@@ -6,26 +6,25 @@ class MultilayerPerceptron(object):
...
                  output_size, name: str = 'multilayer_perceptron',
-                 activation_fn=tf.nn.relu) -> None:
+                 activation_fn=tf.nn.relu, train_mode: tf.Tensor=None) -> None:
```
Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: train_mode by nikdy nemělo bejt None.. Za to ti možná už vynadalo mypy. Musí to být povinný parametr

File: neuralmonkey/nn/mlp.py [hunk 5]
```diff
@@ -6,26 +6,25 @@ class MultilayerPerceptron(object):
...
             self.n_params = 0
             for size in layer_configuration:
                 self.n_params += last_layer_size * size
                 last_layer_size = size
 
-            with tf.variable_scope("classification_layer"):
+            with tf.variable_scope("classification_layer") as varscope:
                 self.n_params += last_layer_size * output_size
-                self.logits = linear(last_layer, output_size)
+                self.logits = linear(last_layer, output_size, scope=varscope)
```
Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: Tohle imo rozsype zpětnou kompatibilitu modelů, v týhle verzi bych to nedělal.

File: neuralmonkey/nn/projection.py [hunk 6]
```diff
@@ -18,15 +19,15 @@ def linear(inputs, size, scope="LinearProjection"):
...
+    with tf.variable_scope(scope) as varscope:
         if isinstance(inputs, list):
             # if there is a list of tensor on the input, concatenate along
             # the last dimension and project.
             inputs = tf.concat(inputs, axis=-1)
 
         return tf.contrib.layers.fully_connected(
             inputs, size, biases_initializer=tf.zeros_initializer(),
-            activation_fn=None, scope=scope)
+            activation_fn=None, scope=varscope)
```
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: nejsem si jistej, jestli tohle náhodou taky nerozseká zpětnou kompatibilitu modelů. Spíš bych se na to vykašlal dělat tuhle změnu.

File: neuralmonkey/nn/projection.py [hunk 7]
```diff
@@ -63,8 +64,8 @@ def maxout(inputs, size, scope="MaxoutProjection"):
     Returns:
         A tensor of shape batch x size
     """
-    with tf.variable_scope(scope):
-        projected = linear(inputs, size * 2, scope=scope)
+    with tf.variable_scope(scope) as varscope:
+        projected = linear(inputs, size * 2, scope=varscope)
```
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: ditto tady

File: neuralmonkey/nn/projection.py [hunk 8]
```diff
@@ -74,14 +75,15 @@ def maxout(inputs, size, scope="MaxoutProjection"):
 
 
 def multilayer_projection(input_, layer_sizes, activation=tf.nn.relu,
-                          dropout_plc=None, scope="mlp"):
+                          dropout_keep_prob=None, train_mode: tf.Tensor=None,
```
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: Zase, train_mode sice tady může bejt `None`, ale nepsal bych to takhle. Dal bych `dropout_keep_prob` defaultně na `1.0`, tak jak je to i jinde a train_mode jako povinný parametr, a pak pryč tu podmínku na řádku 86

File: neuralmonkey/nn/projection.py [hunk 9]
```diff
@@ -74,14 +75,15 @@ def maxout(inputs, size, scope="MaxoutProjection"):
...
+                          scope="mlp"):
     mlp_input = input_
 
     with tf.variable_scope(scope):
         for i, size in enumerate(layer_sizes):
             mlp_input = nonlinear(mlp_input, size, activation=activation,
                                   scope="mlp_layer_{}".format(i))
-            if dropout_plc is not None:
-                mlp_input = tf.nn.dropout(mlp_input, dropout_plc)
+            if dropout_keep_prob is not None:
```
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: tadytu podmínku pryč, funkce dropout když dostane keep_prob jedna, tak jen předá dál to, co dostala, takže už se to tam čekuje

File: neuralmonkey/nn/utils.py [hunk 10]
```diff
@@ -15,12 +16,9 @@ def dropout(variable: tf.Tensor,
         train_mode: A bool Tensor specifying whether to dropout or not
     """
     # Maintain clean graph - no dropout op when there is none applied
-    # TODO maybe use math.isclose instead of this comparison
-    with tf.variable_scope("dropout"):
+    with tf.variable_scope(scope):
```
Comment (jindrahelcl) [inline neuralmonkey/nn/utils.py]: tady vůbec nemá co dělat `variable_scope`, protože se tu nevyrábí žádný variables. Celý `scope` parametr můžeš dát do háje a nezmění se tim nic.

**Comment by jindrahelcl** [comment]
Používá-li se teď ta funkce dropout úplně všude, musí se k ní ještě udělat unit testy. Tim, že je oddělená od zbytku grafu, tak to ani nebude to moc práce.

**Comment by tomkocmi** [comment]
bohuzel netusim jak se delaji unit testy. Udelal bys je prosim? Idealne bych tuhle vetev chtel mergnout do mastra v sobotu pokud projdou reviews, abych mohl v multitasku hledat lepe chyby (mam to tam taky na plno mistech, ale nemyslim si ze by tohle odstranillo chybu, spis az budu svoje dropouty prepisovat tak to muze nekde neco objevit)

**Comment by jindrahelcl** [comment]
Můžu ho tam připravit, ale bylo by dobrý, kdybys tam pak přidal use casy. Jednak se naučíš dělat unit testy do vopice (fakt to neni složitý) a jednak je to šlechetná práce, kterou si všichni zaslouží dělat rovným dílem. :-)

**Comment by tomkocmi** [comment]
Tak kdyz mi napises/ukazes jak tak to klidne udelam, ale doted jsem nedelal ani jeden :D

**Comment by jindrahelcl** [comment]
Udělej si `git pull --rebase origin cleanup` a podívej se.. Něco v takovýmhle stylu, aby to ošetřilo všechny možný divnosti, co ten dropout může dostat na vstup.

**Comment by jindrahelcl** [comment]
všechny ty různý assert funkce jsou zdokumentovaný tady: https://docs.python.org/3/library/unittest.html

File: neuralmonkey/tests/test_nn_utils.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+        for kprob in [0.1, 0.7]:
+            dropped_var = dropout(var, kprob, tf.constant(True))
+            dropped_size = tf.reduce_sum(tf.to_int32(tf.equal(dropped_var, 0.0)))
+
+            dsize = s.run(dropped_size)
+
+            expected_dropped_size = 10000 * (1 - kprob)
+
+            self.assertTrue(np.isclose(expected_dropped_size, dsize, atol=100))
```
Comment (jlibovicky) [inline neuralmonkey/tests/test_nn_utils.py]: Ještě chybí test, že když je `train_mode` `False`, tak se žádný dropout nestane.

File: neuralmonkey/nn/utils.py [hunk 12]
```diff
@@ -15,12 +15,8 @@ def dropout(variable: tf.Tensor,
         train_mode: A bool Tensor specifying whether to dropout or not
     """
     # Maintain clean graph - no dropout op when there is none applied
-    # TODO maybe use math.isclose instead of this comparison
-    with tf.variable_scope("dropout"):
```
Comment (jlibovicky) [inline neuralmonkey/nn/utils.py]: Proč se dává pryč ten scope?

File: neuralmonkey/nn/mlp.py [hunk 13]
```diff
@@ -1,29 +1,31 @@
...
+                mlp_input, layer_configuration, activation=activation_fn,
+                dropout_keep_prob=dropout_keep_prob, train_mode=train_mode,
+                scope="deep_output_mlp")
             self.n_params = 0
             for size in layer_configuration:
                 self.n_params += last_layer_size * size
                 last_layer_size = size
 
-            with tf.variable_scope("classification_layer"):
+            with tf.variable_scope("classification_layer") as varscope:
```
Comment (jlibovicky) [inline neuralmonkey/nn/mlp.py]: K čemu je tady ten pojmenovaný scope, když se to jeho jméno nikde nepoužívá?

File: neuralmonkey/nn/projection.py [hunk 14]
```diff
@@ -18,7 +19,7 @@ def linear(inputs, size, scope="LinearProjection"):
     Returns:
         A tensor of shape batch x size
     """
-    with tf.variable_scope(scope):
+    with tf.variable_scope(scope) as varscope:
```
Comment (jlibovicky) [inline neuralmonkey/nn/projection.py]: A vůbec všude...

File: neuralmonkey/nn/utils.py (hunk 12 above)
Comment (jindrahelcl) [inline neuralmonkey/nn/utils.py]: protože je tam k ničemu, žádný proměnný se v něm nevytvářej

File: neuralmonkey/nn/mlp.py (hunk 13 above)
Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: to je pozůstatek z doby před mym review.

File: neuralmonkey/nn/utils.py (hunk 12 above)
Comment (jlibovicky) [inline neuralmonkey/nn/utils.py]: Jasně. V tom případě bych udělal `namescope`, ten `tf.where` je v tensorboardu docela nepřehlednej.

File: neuralmonkey/nn/mlp.py (hunk 13 above)
Comment (jlibovicky) [inline neuralmonkey/nn/mlp.py]: Aspoň, že to hlásí ten pylint. :-)

File: neuralmonkey/nn/utils.py (hunk 12 above)
Comment (jindrahelcl) [inline neuralmonkey/nn/utils.py]: To by šlo, ale nerval bych to jako parametr, ale nastavil bych mu napevno "dropout"

File: neuralmonkey/nn/mlp.py (hunk 13 above)
Comment (tomkocmi) [inline neuralmonkey/nn/mlp.py]: ja to tam chtel nechat umyslne aby se na to pri pristim refaktoru nezapomenulo, ach jo :( ... tak ja to odeberu.

Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: při příštím refaktoru by to nemělo být potřeba

Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: btw. udělej si cvičení z gitu - podle mě stačí revertovat komit 99701d2781a76184e54ab33d01955ca56d46d153

Comment (tomkocmi) [inline neuralmonkey/nn/mlp.py]: no to bych neriskoval protoze zlikviduju zase celou vetev jako obvykle ... jinak bylo by to pri refaktoru uzitecny, koukni se do soucasne podoby grafu kde je plno krabicek vnorenych nekolikrat do hloubky a je to tim neprehledny

File: neuralmonkey/tests/test_nn_utils.py [hunk 15]
```diff
@@ -34,6 +34,18 @@ def test_keep_prob(self):
             expected_dropped_size = 10000 * (1 - kprob)
 
             self.assertTrue(np.isclose(expected_dropped_size, dsize, atol=100))
+            
+    def test_train_false(self):
+        """Checks that dropout does not """
```
Comment (jindrahelcl) [inline neuralmonkey/tests/test_nn_utils.py]: does not co? :D

Comment (tomkocmi) [inline neuralmonkey/tests/test_nn_utils.py]: sakra jsem zacal resit jak spravne udelat unit test a nedopsal jsem to :D ... tak pak to pridam k poslednimu commitu, ale mimo to by to takhle mohlo uz byt?

**Comment by tomkocmi** [comment]
ping (ten test opravim v poslednim commitu at to neplevelim)

File: neuralmonkey/nn/mlp.py [hunk 16]
```diff
@@ -6,18 +7,19 @@ class MultilayerPerceptron(object):
...
 
     # pylint: disable=too-many-arguments
-    def __init__(self, mlp_input, layer_configuration, dropout_plc,
-                 output_size, name: str = 'multilayer_perceptron',
-                 activation_fn=tf.nn.relu) -> None:
+    def __init__(self, mlp_input: tf.Tensor, layer_configuration: List[int],
+                 dropout_keep_prob: float,
+                 output_size: int, train_mode: tf.Tensor,
+                 name: str = 'multilayer_perceptron',
+                 activation_fn=tf.nn.relu, ) -> None:
```
Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: tady přidat anotaci k activation_fn.. Bude to  něco jako `Callable[[tf.Tensor], tf.Tensor]`

Comment (jindrahelcl) [inline neuralmonkey/nn/mlp.py]: a taky to hoď argument na řádku, jako je to na většině místech, který maj `too-many-arguments`...

File: neuralmonkey/nn/projection.py [hunk 17]
```diff
@@ -73,15 +74,15 @@ def maxout(inputs, size, scope="MaxoutProjection"):
         return reshaped
 
 
-def multilayer_projection(input_, layer_sizes, activation=tf.nn.relu,
-                          dropout_plc=None, scope="mlp"):
+def multilayer_projection(input_, layer_sizes, train_mode: tf.Tensor,
+                          activation=tf.nn.relu, dropout_keep_prob=1.0,
+                          scope="mlp"):
```
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: dodělat anotace (tensor, pole intů, tensor, callable z tensoru na tensor, float, string)

**Comment by jindrahelcl** [comment]
btw. (@jlibovicky, ale i @tomkocmi) podle mě není dobrá praxe dávat approve do zálohy před opravením testů.. Oprava testu může přinýst bugy úplně stejně jako jakákoliv jiná úprava kódu a měla by se stejně tak reviewovat.

**Comment by tomkocmi** [comment]
Tak vy jste mi rikali, ze nechcete commity co opravujou pylintovy hlasky, takze proto vzdycky cekam abych ty veci s pylintem pridal k jinemu commitu. Pokud to vsak uz neplati, tak rad vse pripravim pred approvem.

**Comment by jindrahelcl** [comment]
Tomu nevěřim, že jsme to řikali. Sami to děláme. Spíš jsme někdy řikali, žes udělal deset komitů za sebou, všechny týkající se opravy chyb, co travis nahlásil, když to mohl být jeden.

**Comment by tomkocmi** [comment]
proc byla posledni oprava nutna? Nechapu kde maji a kde nemaji byt mezery.

activation: Callable[[tf.Tensor], tf.Tensor]=tf.nn.relu,
dropout_keep_prob: float = 1.0,

proc v prvnim pripade nesmi byt mezery okolo rovnitka a v druhem musi? Vzdyt to nedava smysl

**Comment by jindrahelcl** [comment]
@tomkocmi to je bug v pylintu, na kterej už je tu někde odkaz. Všude, kde je v názvu typu tečka (např. `tf.Tensor`), tak tam mezera sice být má, ale pylint na tom zahlásí error. Takže zatím naše politika je vyhovět v těchto případech špatnýmu pylintu, a až se aktualizuje verze se zafixovaným bugem (horizont několika týdnů, max měsíců), tak to všechno začne padat a všechno se to v jednom komitu opraví.

File: neuralmonkey/encoders/cnn_encoder.py [hunk 18]
```diff
@@ -141,7 +139,8 @@ def __init__(self,
                 self.encoded = multilayer_projection(
                     last_layer_flat, fully_connected,
                     activation=tf.nn.relu,
-                    dropout_plc=self.dropout_placeholder)
+                    dropout_keep_prob=dropout_keep_prob,
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/cnn_encoder.py]: tady by asi mělo být spíš `self.dropout_keep_prob`, když se to tam na začátku přiřazuje. Teoreticky kdyby se volala nějaká funkce, která mění `self.dropout_keep_prob`, tak by se to tady neprojevovalo.
